src/pages.rs

```diff
@@ -35,13 +37,70 @@ impl From<crate::Error> for ErrorResponse<'_> {
 pub struct Index<'a> {
     meta: &'a env::Metadata<'a>,
     base_path: &'static env::BasePath,
+    max_expiry: Option<u32>,
+
+    /// SAFETY: calls in the template are always sequential
+    default_has_been_written: Cell<bool>,
 }
 
 impl<'a> Default for Index<'a> {
     fn default() -> Self {
         Self {
             meta: env::metadata(),
             base_path: env::base_path(),
+            max_expiry: env::max_paste_expiry(),
+            default_has_been_written: Cell::new(false),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Expiry<'a> {
+    Special(&'a str),
+    Time(u32),
+}
+
+impl<'a> Expiry<'a> {
+    fn as_str(&self) -> String {
+        match self {
+            Expiry::Special(s) => s.to_string(),
+            Expiry::Time(u) => u.to_string(),
+        }
+    }
+}
+
+impl<'a> Index<'a> {
+    fn expiry(&self, name: &str, time: Expiry<'a>) -> String {
+        let sel_string = if self.default_has_been_written.get() {
+            ""
+        } else {
+            r#" selected"#
+        };
+
+        match self.max_expiry {
+            Some(exp) => {
+                match time {
+                    // never emit an never expire with a limit
+                    Expiry::Special("") => String::new(),
+                    Expiry::Special("burn") => {
+                        self.default_has_been_written.set(true);
+                        format!(r#"<option{} value="burn">{}</option>"#, sel_string, name)
+                    }
```

Comment:
I guess you could further discriminate `Expiry::Special("")`, `Expiry::Special("burn")` and `Expiry::Special(_)` into their own enum variants? Seems a bit half full, half empty :-)

Reply:
Yeea, i kinda wrote this at a deep in the night kinda time. Depending on if we move the whole generation into a function id completely remove the enum, or make these proper variants.